storage/remote/wal_watcher.go

```diff
@@ -0,0 +1,247 @@
+		// case <-w.ctx.Done():
+		// 	return
+		case event, ok := <-watcher.Events:
+			_, fileName := filepath.Split(event.Name)
+			// Check if the file name is a new segment and not a checkpoint or something else.
+			if ok := filenameRegex.MatchString(fileName); !ok {
+				w.logger.Log("eventName", event.Name, "msg", "did not match regex")
+				continue
+			}
+			w.logger.Log("msg", "watcher event")
+			if !ok {
+				return
+			}
+			if event.Op&fsnotify.Write == fsnotify.Write {
+				if fileName != currentSegment {
+					continue
+				}
+				w.logger.Log("event", "write", "file", event.Name)
+				w.decodeSegment(segment)
+			}
+			if event.Op&fsnotify.Create == fsnotify.Create {
+				// TODO: callum, ensure the create was for the segment n+1 and not n+2 or more
+				w.logger.Log("event", "create", "file", event.Name)
+				// Read the current segment one last time in case we receive events out of order. Not sure if this is necessary.
+				w.decodeSegment(segment)
+				segment.Close()
+				n++
+				// Reset series map, each Segment file keeps track of series independently and there may be series churn over time.
+				w.series = make(map[uint64][]prompb.Label)
+				segment, err = wal.OpenReadSegment(wal.SegmentName(w.walDir, n))
+				if err != nil {
+					w.logger.Log("err", err)
+					return
+				}
+				currentSegment = fmt.Sprintf("%08d", n)
+			}
+		case err, ok := <-watcher.Errors:
+			if !ok {
+				return
+			}
+			level.Error(w.logger).Log("err", err)
+		}
+	}
+}
+
+func (w *WALWatcher) decodeSegment(segment *wal.Segment) bool {
+	// todo: callum, is there an easy way to detect if r.Next() has returned the last possible record in a segment.
+	// For now just recover from the panic that is thrown if we call r.Next() when we shouldn't have
+	// defer func() {
+	// 	if r := recover(); r != nil {
+	// 		level.Error(log.With(w.logger)).Log("err", r)
+	// 		segment.Close()
+	// 	}
+	// }()
+	r := wal.NewReader(segment)
+	var (
+		dec      tsdb.RecordDecoder
+		series   []tsdb.RefSeries
+		samples  []tsdb.RefSample
+		nSeries  = 0
+		nSamples = 0
+	)
+	for {
+		ok := r.Next()
+		rec := r.Record()
+		switch dec.Type(rec) {
+		case tsdb.RecordSeries:
+			w.logger.Log("msg", "new series")
+			series, err := dec.Series(rec, series[:0])
+			if err != nil {
+				level.Error(log.With(w.logger)).Log("err", err)
+				return ok
+			}
+			for _, q := range w.storage.queues {
+				q.storeSeries(series)
+			}
+			nSeries += len(series)
+		case tsdb.RecordSamples:
+			w.logger.Log("msg", "new samples")
+			samples, err := dec.Samples(rec, samples[:0])
+			if err != nil {
+				level.Error(log.With(w.logger)).Log("err", err)
+				return ok
+			}
+
+			for _, q := range w.storage.queues {
+				for _, s := range samples {
+					q.Append(s)
+				}
+			}
+			nSamples += len(samples)
+		case tsdb.RecordTombstones:
+			w.logger.Log("msg", "got tsdb tombstones")
+		case tsdb.RecordInvalid:
```

Comment:
@gouthamve or @fabxc any pointers on what would be causing us to read back invalid records? Should I be worried that we're reading back invalid/corrupt or incomplete data?